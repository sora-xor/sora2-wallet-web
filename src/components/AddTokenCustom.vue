<template>
  <div class="token-custom">
    <s-input
      class="token-custom-address"
      :maxlength="128"
      :placeholder="t(`addToken.${AddTokenTabs.Custom}.addressPlaceholder`)"
      v-model="address"
      @change="handleSearch"
    />
    <s-input
      class="token-custom-symbol"
      :maxlength="10"
      :placeholder="t(`addToken.${AddTokenTabs.Custom}.symbolPlaceholder`)"
      v-model="symbol"
    />
    <div class="token-custom-actions s-flex">
      <s-button @click="handleBack">{{ t('addToken.cancel') }}</s-button>
      <s-button type="primary" :disabled="!(symbol && address)" @click="handleAddToken">
        {{ t('addToken.action') }}
      </s-button>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from 'vue-property-decorator'
import { Action } from 'vuex-class'

import TranslationMixin from './mixins/TranslationMixin'
import { AddTokenTabs, RouteNames } from '../consts'
import { getTokenIconClasses } from '../util'

@Component
export default class AddTokenCustom extends Mixins(TranslationMixin) {
  readonly AddTokenTabs = AddTokenTabs

  @Action navigate
  @Action searchToken
  @Action addToken

  address = ''
  symbol = ''
  selectedToken: any = null

  async handleSearch (value: string): Promise<void> {
    if (!value.trim()) {
      this.selectedToken = null
      this.symbol = ''
      return
    }
    const search = value.trim().toLowerCase()
    // TODO: there is shouldn't be an already attached token
    // For now it's just for tests
    this.selectedToken = await this.searchToken({ search })
    if (this.selectedToken && this.selectedToken.symbol) {
      this.symbol = this.selectedToken.symbol
    } else {
      this.selectedToken = null
      this.symbol = ''
    }
  }

  handleSelectToken (token: any): void {
    this.selectedToken = token
  }

  handleBack (): void {
    this.navigate({ name: RouteNames.Wallet })
  }

  async handleAddToken (): Promise<void> {
    const token = { symbol: this.symbol, address: this.address }
    await this.addToken({ token })
    this.navigate({ name: RouteNames.Wallet })
    this.$emit('add-token')
  }
}
</script>

<style scoped lang="scss">
.token-custom {
  > * {
    margin-top: $basic-spacing;
  }
  &-actions {
    > * {
      flex: .5;
    }
  }
}
</style>
