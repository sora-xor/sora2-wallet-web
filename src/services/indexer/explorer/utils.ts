import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
