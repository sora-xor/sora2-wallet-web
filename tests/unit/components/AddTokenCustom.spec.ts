import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'

import AddTokenCustom from '@/components/AddTokenCustom.vue'
import { TranslationMock, SoramitsuElementsImport } from '../../utils'

const localVue = createLocalVue()
localVue.use(Vuex)
const store = new Vuex.Store({
  modules: {
    Router: {
      actions: {
        navigate: jest.fn()
      }
    },
    Account: {
      actions: {
        searchToken: jest.fn(),
        addToken: jest.fn()
      }
    }
  } as any
})

describe('AddTokenCustom.vue', () => {
  beforeEach(() => {
    SoramitsuElementsImport(localVue)
    TranslationMock(AddTokenCustom)
  })

  it('should renders correctly', () => {
    const wrapper = shallowMount(AddTokenCustom, { localVue, store })
    expect(wrapper.element).toMatchSnapshot()
  })
})
