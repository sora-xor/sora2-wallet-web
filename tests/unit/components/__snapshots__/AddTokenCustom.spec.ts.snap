// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTokenCustom.vue should renders correctly 1`] = `
<div
  class="token-custom"
>
  <s-input-stub
    accept=""
    autocomplete="off"
    borderradius="mini"
    class="token-custom-address"
    form=""
    label=""
    maxlength="128"
    name=""
    placeholder=""
    prefix=""
    size="big"
    suffix=""
    tabindex=""
    type="text"
    value=""
  />
   
  <s-input-stub
    accept=""
    autocomplete="off"
    borderradius="mini"
    class="token-custom-symbol"
    form=""
    label=""
    maxlength="10"
    name=""
    placeholder=""
    prefix=""
    size="big"
    suffix=""
    tabindex=""
    type="text"
    value=""
  />
   
  <div
    class="token-custom-actions s-flex"
  >
    <s-button-stub
      borderradius="small"
      icon=""
      iconposition="left"
      nativetype="button"
      size="medium"
      tabindex="0"
      tooltip=""
      tooltipplacement="top"
      type="secondary"
    >
      
    </s-button-stub>
     
    <s-button-stub
      borderradius="small"
      disabled="true"
      icon=""
      iconposition="left"
      nativetype="button"
      size="medium"
      tabindex="0"
      tooltip=""
      tooltipplacement="top"
      type="primary"
    >
      
      
    
    </s-button-stub>
  </div>
</div>
`;
